drivers/timer/hi3861_timer.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2023 Chen Xingyu <hi@xingrz.me>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT hisilicon_hi3861_timer
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/timer/system_timer.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/irq.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/sys_clock.h>
+#include <zephyr/arch/common/sys_io.h>
+
+#define TIMER_REG DT_INST_REG_ADDR(0)
+
+#define TIMER_0 (TIMER_REG + 0x00)
+#define TIMER_1 (TIMER_REG + 0x14)
+#define TIMER_2 (TIMER_REG + 0x28)
+#define TIMER_3 (TIMER_REG + 0x3C)
+
+#define TIMER_LOADCOUNT_REG(BASE)    (BASE + 0x000)
+#define TIMER_CURRENTVALUE_REG(BASE) (BASE + 0x004)
+#define TIMER_CONTROLREG_REG(BASE)   (BASE + 0x008)
+#define TIMER_CONTROLREG_EN          (1 << 0)
+#define TIMER_CONTROLREG_MODE_FREE   (0 << 1)
+#define TIMER_CONTROLREG_MODE_CYCLE  (1 << 1)
+#define TIMER_CONTROLREG_INT_MASK    (1 << 2)
+#define TIMER_CONTROLREG_INT_UNMASK  (0 << 2)
+#define TIMER_CONTROLREG_LOCK        (1 << 3)
+#define TIMER_EOI_REG(BASE)          (BASE + 0x00C)
+#define TIMER_INTSTATUS_REG(BASE)    (BASE + 0x010)
+
+#define SYSTICK TIMER_3
+
+#define CYCLES_PER_TICK (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+#if defined(CONFIG_TEST)
+const int32_t z_sys_timer_irq_for_test = DT_INST_IRQN(0);
+#endif
+
+static struct k_spinlock lock;
+
+static volatile uint32_t timeout_ticks;
+static volatile uint64_t announced_ticks;
+
+static inline void timer_setup(uint32_t cycles)
+{
+	sys_write32(cycles, TIMER_LOADCOUNT_REG(SYSTICK));
+	sys_write32(TIMER_CONTROLREG_MODE_CYCLE | TIMER_CONTROLREG_INT_UNMASK,
+		    TIMER_CONTROLREG_REG(SYSTICK));
+}
+
+static inline void timer_stop(void)
+{
+	sys_write32(0, TIMER_CONTROLREG_REG(SYSTICK));
+}
+
+static inline void timer_start(void)
+{
+	uint32_t regval;
+
+	regval = sys_read32(TIMER_CONTROLREG_REG(SYSTICK));
+	regval |= TIMER_CONTROLREG_EN;
+
+	sys_write32(regval, TIMER_CONTROLREG_REG(SYSTICK));
+}
+
+static void sys_clock_isr(const void *arg)
+{
+	ARG_UNUSED(arg);
+
+	uint32_t dticks;
+
+	/* Read EOI to clear interrupt */
+	sys_read32(TIMER_EOI_REG(SYSTICK));
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+#if defined(CONFIG_TICKLESS_KERNEL)
+	timer_stop();
```

Comment:
why you need to stop the driver? You cannot stop the time form flowing right?

Reply:
The systick timer on this SoC has two modes: free mode and period mode. In free mode, it operates as a one-shot timer, counting down from `0xFFFFFFFF` to zero. In period mode, it counts down from the value programmed in the `LOADCOUNT` register to zero and then repeats the process from `LOADCOUNT`. As a result, since the count cannot be programmed in free one-shot mode, I can only implement the driver using the period mode and disable it once it has been triggered.

Reply:
when operating in free mode, do you have any compare register to use? Or it just fires the interrupt when expiring to 0? In that case how you can possibly use that as timer then?

Reply:
No, there is no comparing function available on this timer. This also reminds me of a new problem - I cannot determine exactly how many ticks have elapsed after an ISR is triggered and before the next timeout is set.

Reply:
@andyross @teburd do you have any other driver like this one in tree to see how this is implemented there?

Reply:
Anyway, is it okay to implement just the tickful mode? Just like `leon_gptimer.c`.

Reply:
it is, especially if fully tickless is not achievable.